### DISCUSSION TITLE
Unclear what the column 'E' is outputting in the console output during training

### DISCUSSION THREAD

**Discussion by erotavlas** [explosion/spaCy#7731] [Help: Other Questions] [answered]
At first I thought it was an iteration within each Epoch but now I'm not sure after seeing it  complete an entire training.  I set my max epochs parameter in the configuration to 25 and this was the result.  It looks like there are 25 rows but the Epoch number doesn't correspond to the max value I specified in the configuration.  Also I'm not sure how this related to the 'iterations' parameter that we used to use in the spaCy 2.x. Thank s for any clarification.

![image](https://user-images.githubusercontent.com/10454265/114210666-e933db80-992d-11eb-9bbc-698040b58f88.png)

**Comment by polm** [comment, marked as answer]
E is for epochs, # is for steps (iterations). Training can end before the max epochs / steps if you've set a `patience` value. You can read more about how these values interact in https://github.com/explosion/spaCy/discussions/7465. 

It looks like your score (right column) didn't improve for 8 * 200 = 1600 steps, which is the default patience value, and should be listed in your config. 

Does that make sense?

**Reply by erotavlas** [reply]
It's kind of confusing.  Maybe you should relabel the # column with the actual meaning (i.e. Steps)
I'm not clear what Steps are, but in TensorFlow, patience is based on number of epochs, not steps which also contributes to the confusion here.

**Reply by DavidGerva** [reply]
@polm (or maybe also @erotavlas) I continue here because I think can be useful for someone else.

I still don't get what a "**step**" is: In one epoch the earning algorithm works through the entire training dataset once. Right? But what does the algorithm (spacy) is doing in one step?

**Reply by adrianeboyd** [reply]
A step is running `nlp.update` for one training batch, with a batch as configured by `training.batcher`.

**Reply by kumarsuraj9450** [reply]
I'm still not getting the whole process. Suppose if i have 20k documents in training data and i want to do 5 epochs what should be the configuration for batch size, eval frequency and max steps If patience is set to 0?


Also correct me if I'm wrong but a step means one batch is processed. And one batch means "n" number of data points from training dataset. Like in above code a 100 datapoints from training dataset. 

Weights gets updated after each batch or step right?

**Reply by adrianeboyd** [reply]
To disable everything else and use only `max_epochs`, set `max_steps` and `patience` to `0`.

Batch size and eval frequency are independent of the the number of epochs.

Batch sizes aren't necessarily static, they are determined by the `[training.batcher]` settings in the config.

Eval frequency mainly matters in terms of how often a `model-best` model is saved and for `patience` (early stopping). In terms of what is saved as the `model-last` model, eval frequency doesn't matter. For training for a fixed epoch count, eval frequency could be very high to reduce time spent on eval during training.

Weights are updated after each batch/step.

Please also take a look at the discussion linked above: #7465
